src/cart/crt_iv.c

```diff
@@ -2706,6 +2924,12 @@ bulk_update_transfer_done_aux(const struct crt_bulk_cb_info *info)
 	update_cb_info->uci_bulk_hdl = cb_info->buc_bulk_hdl;
 
 	if (update_rc == -DER_IVCB_FORWARD) {
+		/* Forward request to the parent    */
+
+		/* Get group version to associate with next_rank. */
+		/* Pass it down to crt_ivu_rpc_issue	      */
+		grp_ver = ivns_internal->cii_grp_priv->gp_membs_ver;
+
```

Comment:
there is a problem here. this grp_ver we retrieve here has to correspond to 'next_rank' that we get on following line below. As such we need to lock ivns_intenral->cii_grp_priv so that version and rank we retrieve (based on groups membership) are in sync with each other

Reply:
No,  If there is no changed in group version andby the time we get "next"rank", then they are in sync.

If the group version has changed, then they will be out of sync.  However, this group version is passed down to crt_ivu_rpc_issue which does another check prior building the rpc request, at which time the problem will be detected.

Previously, crt_ivu_rpc_issue was using the local group version; which was in error.  If the group version changed from the time when "next_rank" was found and the construction of the rpc structure, then the rpc structure was not correct.  It must used the version passed to it.

Doing it this way eliminates any requirements of using a semaphore; which simplifies the code.  If a semaphore was used, then it would have to have it the entire time of building the rpc structure.  Plus, the semaphore usage would have to be implemented in all code that checks for version.  This method works even if some other code wishes to use a semaphore around version checking.

I could have also compared the group version before and after retrieving "next_rank" and exited out early.  But the version would still have to be passed down to crt_ivu_issue_rpc.

Reply:
If you use group lock then you only need to keep it during query of group version along with the corresponding 'next_rank' calculation. Once those 2 pieces are found, you no longer need to have group locked, as you would be passing proper version into crt_ivu_issue_rpc. 

Thinking on it more, you actually dont even need 2 checks and neither you need a lock. Instead you can simplify this by doing:

First get a 'next_rank'. Next check that group version has not changed. if group version has not changed then just pass queried version into crt_ivu_issue_rpc and use it to populate input when transferring data to 'next_rank'. if there is a group version mismatch then just return error

Reply:
Yes, you could resolve the issue by taking a semaphore. But then you would need to make sure that any code modifying the version numbers also uses the semaphores.  I am trying not to use a semaphore if possible.  I find that it creates a choke point and added complexity.

You need to obtain the first copy of the group version prior to getting the rank.  Otherwise, the group version could change between the getting of the rank and reading the version.  After getting the rank, I could do a check to see if the version has changed.  Rather than that, I put the check in the building of the rpc structure.  The puts the check in at the latest possible time and will more likely catch any change in version number.  crt_ivu_issue_rpc does use the version number passed to it to populate the structure.

Reply:
It is already set that any code that modifies version number or group membership takes group lock.
You are correct regarding group version potentially changing between getting rank and reading version, so in this case yes we still need to protect then both of those things with group lock.

We really do not care about catching group version changing after we've verified that group version matches incoming group version and we've calculated corresponding 'next_rank' using the proper version

Reply:
You do not need to use the semaphore if you do a read-version, read-rank, version-check.  However, if you want the semaphore, then I will implement that.

Reply:
yes, but i think it makes for cleaner code if we just do:

D_RWLOCK_RDLOCK(&grp_priv->gp_rwlock);
read version
read rank
D_RWLOCK_UNLOCK(&grp_priv->gp_rwlock);
compare version and fail or continue; pass read version to followup calls

this way the check is in a single place instead of being split across multiple calls

Reply:
Locks were implemented.